packages/compose/src/hooks/use-observe-element-size/index.ts

```diff
@@ -0,0 +1,109 @@
+/**
+ * WordPress dependencies
+ */
+import { useRef, useEffect } from '@wordpress/element';
+/**
+ * Internal dependencies
+ */
+import useEvent from '../use-event';
+
+/**
+ * Tracks a given element's size and calls `onUpdate` for all of its discrete
+ * values using a `ResizeObserver`. The element can change dynamically and **it
+ * must not be stored in a ref**. Instead, it should be stored in a React
+ * state or equivalent.
+ *
+ * @param targetElement         The target element to observe. It can be changed dynamically.
+ * @param onUpdate              Callback that will be called when the element is resized. It is passed the list of [`ResizeObserverEntry`](https://developer.mozilla.org/en-US/docs/Web/API/ResizeObserverEntry) objects passed to the `ResizeObserver.observe` callback internally, and the element being tracked at the time of this update.
+ * @param resizeObserverOptions Options to pass to `ResizeObserver.observe` when called internally. Updating this option will not cause the observer to be re-created, and it will only take effect if a new element is observed.
+ *
+ * @example
+ *
+ * ```tsx
+ * const [ targetElement, setTargetElement ] = useState< HTMLElement | null >();
+ * useObserveElementSize(
+ * 	targetElement,
+ * 	( resizeObserverEntries, element ) => {
+ * 		console.log( 'Resize observer entries:', resizeObserverEntries );
+ * 		console.log( 'Element that was resized:', element );
+ * 	},
+ * 	{ box: 'border-box' }
+ * );
+ * <div ref={ setTargetElement } />;
+ * ```
+ */
+export default function useObserveElementSize(
+	/**
+	 * The target element to observe. It can be changed dynamically.
+	 */
+	targetElement: HTMLElement | undefined | null,
+	/**
+	 * Callback that will be called when the element is resized.
+	 */
+	onUpdate: (
+		/**
+		 * The list of
+		 * [`ResizeObserverEntry`](https://developer.mozilla.org/en-US/docs/Web/API/ResizeObserverEntry)
+		 * objects passed to the `ResizeObserver.observe` callback internally.
+		 */
+		resizeObserverEntries: ResizeObserverEntry[],
+		/**
+		 * The element being tracked at the time of this update.
+		 */
+		element: HTMLElement
```

Comment:
This is redundant because the element is always in `entry.target` of each of the entries. `ResizeObserver` can track multiple elements, e.g., when `.observe( el )` is called multiple times.

Reply:
It is a nicer API if you don't need the entries. See my utilities for instance (`useTrackElementOffsetRect`). Since it's the last argument, it doesn't hurt and is compatible with usage that actually does need entries.

Reply:
Yes, it doesn't hurt, but still redundant. 🙂 We could reduce the `entries` array to just the last entry (all the others are useless?) and then access the element as:
```js
onUpdate = ( entry ) => {
  const element = entry.target;
};
```

Reply:
Removing some entries breaks users who need those entries. Passing element as an extra argument you can ignore has no effect for those who don't need it. It's just a nicer API. All abstractions are redundant by that logic, as you can always obtain the same data/behavior using the lower level tools :P

If you need to measure the element or do other stuff with it (as I do in my use case where I not only need a specific size information, but also positioning data not provided by entries), `(_, element) => element.getBoundingClientRect()` is way cleaner and easier to understand than `(entries) => entries[0]!.target.getBoundingClientRect()`.

Reply:
Yes, but the `element` parameter is still redundant, and it's there only for convenience in one specific use case. You can do the same thing with `entries.at(-1).target`.

Imagine proposing a new standard DOM API, a JS feature in TC39, or a native API in React. An extra parameter that's there merely for convenience, and doesn't enable any new functionality, would never be accepted. You would be reminded that perfection is achieved only when there is nothing left to take away, and asked to update the proposal. Let's apply the same standard here.

Also, let's try to evolve the hook into having an API like:
```js
const elRef = useObserveElementSize( ( entries ) => { ... }, options );
return <div ref={ elRef } />;
```
The typical way to work with DOM elements is refs. And this notation nicely mirrors how you would use a plain ref:
```js
const elRef = useRef();
return <div ref={ elRef } />;
```
and is also used by other libraries like `floating-ui`:
```js
const { refs } = useFloating();
return <div ref={ refs.setFloating } />;
```

If there is an atypical use case where the elements are not in refs, but, for example, in a data store, then it will have to go through a bit of inconvenience, something to the tune of:
```js
const { element } = store.item( id );
const elementRef = useObserveElementSize();
useEffect( () => { elementRef.current = element; }, [ element ] );
```
As long as it's doable, it is a very good cost to pay for a simplified API for `useObserveElementSize`. It's superficially convenient to have APIs that accept almost anything as input and they will internally figure it out, Lodash or jQuery are full of APIs like this, but in the long term they cause a lot of trouble.

Reply:
> I do see value in this abstraction and would like to discuss it on its own merits instead of disregarding it entirely.

I believe we are discussing it on its own merits since the beginning. We just seem to value the same things differently. I don't like the fact that we don't pass through the native platform callbacks of type `ResizeObserverCallback` as they are, that we are extending them. I see that as an unneccesary monkey patching. What if a future standard adds some second argument to the callback?

I don't see the accessibility/usability benefits of `element` to be big enough to justify that. You already have the element in `entries`. When you're getting the element from `tabStore`, you can have a second instance of it available as a local variable. Then why provide yet another way to get it?